ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -0,0 +1,708 @@
+    this.localLagFromRemoteStore = localLagFromRemoteStore;
+  }
+
+  long getTotalBytesReadFromLocalStore() {
+    return this.totalBytesReadFromLocalStore;
+  }
+
+  void setToken(FindToken token) {
+    // reference assignment is atomic in java but we want to be completely safe. performance is
+    // not important here
+    synchronized (lock) {
+      this.currentToken = token;
+    }
+  }
+
+  void initializeTokens(FindToken token) {
+    synchronized (lock) {
+      this.currentToken = token;
+      this.candidateTokenToPersist = token;
+      this.tokenSafeToPersist = token;
+      this.timeCandidateSetInMs = time.milliseconds();
+    }
+  }
+
+  /*
+   * The token persist logic ensures that a token corresponding to an entry in the store is never persisted in the
+   * replicaTokens file before the entry itself is persisted in the store. This is done as follows. Objects of this
+   * class maintain 3 tokens: tokenSafeToPersist, candidateTokenToPersist and currentToken:
+   *
+   * tokenSafeToPersist: this is the token that we know is safe to be persisted. The associated store entry from the
+   * remote replica is guaranteed to have been persisted by the store.
+   *
+   * candidateTokenToPersist: this is the token that we would like to persist next. We would go ahead with this
+   * only if we know for sure that the associated store entry has been persisted. We ensure safety by maintaining the
+   * time at which candidateTokenToPersist was set and ensuring that tokenSafeToPersist is assigned this value only if
+   * sufficient time has passed since the time we set candidateTokenToPersist.
+   *
+   * currentToken: this is the latest token associated with the latest record obtained from the remote replica.
+   *
+   * tokenSafeToPersist <= candidateTokenToPersist <= currentToken
+   * (Note: when a token gets reset by the remote, the above equation may not hold true immediately after, but it should
+   * eventually hold true.)
+   */
+
+  /**
+   * get the token to persist. Returns either the candidate token if enough time has passed since it was
+   * set, or the last token again.
+   */
+  FindToken getTokenToPersist() {
+    synchronized (lock) {
+      if (time.milliseconds() - timeCandidateSetInMs > tokenPersistIntervalInMs) {
+        // candidateTokenToPersist is now safe to be persisted.
+        tokenSafeToPersist = candidateTokenToPersist;
+      }
+      return tokenSafeToPersist;
+    }
+  }
+
+  void onTokenPersisted() {
+    synchronized (lock) {
+      /* Only update the candidate token if it qualified as the token safe to be persisted in the previous get call.
+       * If not, keep it as it is.
+       */
+      if (tokenSafeToPersist == candidateTokenToPersist) {
+        candidateTokenToPersist = currentToken;
+        timeCandidateSetInMs = time.milliseconds();
+      }
+    }
+  }
+
+  @Override
+  public String toString() {
+    return replicaId.toString();
+  }
+}
+
+final class PartitionInfo {
```

Comment:
My first version was make it in a Separate file/class, but turns out only `ReplicationEngine.java` depends on it, so moved it back to `ReplicationEngine.java`. Any preference?

Reply:
If they are only used inside of `ReplicationEngine`, you could consider making them private static inner classes inside of `ReplicationEngine`

Reply:
yes. Making it private inner. 
A separate class file is create for RemoteReplicaInfo since it's used in both ReplicationEngine.java and ReplicaThraed.java.

Reply:
DiskTokenPersistor depends on `PartitionInfo `. It can't be inner class. Making it an individual component.